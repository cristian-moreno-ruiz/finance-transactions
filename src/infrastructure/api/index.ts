export { TransactionRouter } from 'src/infrastructure/api/transaction-router';
export { TransactionController } from './transaction-controller';
