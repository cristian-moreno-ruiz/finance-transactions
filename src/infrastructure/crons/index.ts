export { TransactionIngestionCron } from 'src/infrastructure/crons/transaction-ingestion-cron';
