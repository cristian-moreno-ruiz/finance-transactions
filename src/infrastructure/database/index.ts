export { mongoInit } from 'src/infrastructure/database/mongo-init';
