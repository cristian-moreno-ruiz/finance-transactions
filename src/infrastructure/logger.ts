export class Logger {
	public static log(message?: any, ...optionalParams: any[]): void {
		console.log(message, optionalParams);
	}
	public static error(message?: any, ...optionalParams: any[]): void {
		console.error(message, optionalParams);
	}
}
