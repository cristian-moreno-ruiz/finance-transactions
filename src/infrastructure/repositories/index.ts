export { MongoTransactionRepository } from 'src/infrastructure/repositories/mongo-transaction-repository';
export { TransactionSchema, MongoTransactionModel } from 'src/infrastructure/repositories/mongo-transaction-schema';
