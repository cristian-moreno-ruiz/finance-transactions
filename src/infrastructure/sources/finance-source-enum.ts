export enum FinanceSourceEnum {
	SAMPLE_SOURCE = 'SAMPLE_SOURCE',
}