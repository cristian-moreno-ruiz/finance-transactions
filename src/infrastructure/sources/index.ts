export { SampleSource } from 'src/infrastructure/sources/sample-source';
export { FinanceSourceEnum } from 'src/infrastructure/sources/finance-source-enum';
