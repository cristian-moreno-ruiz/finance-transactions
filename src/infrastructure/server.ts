import bodyParser from 'body-parser';
import dotenv from 'dotenv';
import express, { Application } from 'express';
import cron from 'node-cron';
import { FinanceService, TransactionIngestionService } from 'src/application';
import { TransactionController, TransactionRouter } from 'src/infrastructure/api';
import { TransactionIngestionCron } from 'src/infrastructure/crons/transaction-ingestion-cron';
import { mongoInit } from 'src/infrastructure/database';
import { MongoTransactionRepository } from 'src/infrastructure/repositories';
import { SampleSource } from 'src/infrastructure/sources';

dotenv.config();

const PORT = process.env.PORT || 3000;
const PROVIDER_URL =
	process.env.PROVIDER_URL || 'https://gist.githubusercontent.com/cristian-moreno-ruiz/40984d50552cfd94e715c3731d85d2e4/raw/f64fdba75307f2f581202d6add89b0aebc3a1c59/sample-finance-source.response.xml';
const CRON_EXPRESSION = process.env.CRON_EXPRESSION || '* * * * *';

const app: Application = express();
app.use(bodyParser.json());

async function start() {
	// Connect to MongoDB
	await mongoInit();

	// Orchestrate dependencies
	const mongoRepository = new MongoTransactionRepository();
	const transactionController = new TransactionController(new FinanceService(mongoRepository));
	const transactionRouter = new TransactionRouter(transactionController);
	const transactionIngestionCron = new TransactionIngestionCron(
		new TransactionIngestionService(mongoRepository, new SampleSource(PROVIDER_URL)),
	);

	// Initialize routes
	app.use('/transactions', transactionRouter.router);

	// Start the API server
	app.listen(PORT, () => {
		console.log(`Server running on http://localhost:${PORT}`);
	});

	// Trigger ingestion of transactions from the sample source on startup
	await transactionIngestionCron.triggerTransactionRefresh();

	// Refresh transactions every minute
	cron.schedule(CRON_EXPRESSION, async () => {
		await transactionIngestionCron.triggerTransactionRefresh();
	});
}

start();
