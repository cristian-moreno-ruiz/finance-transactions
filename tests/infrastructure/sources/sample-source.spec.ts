import axios from 'axios';
import fs from 'fs';
import path from 'path';
import { SampleSource } from 'src/infrastructure/sources';
import { v4 as uuidv4 } from 'uuid';

const xmlFilePath = path.resolve(
	__dirname,
	'tests/../../../__fixtures__/sample-finance-source.response.xml',
);
const xmlResponse = fs.readFileSync(xmlFilePath, 'utf-8');
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;
const providerUrl = 'http://test-url.com';

// mock uuid
jest.mock('uuid', () => ({ v4: jest.fn() }));
const mockedUuid = uuidv4 as jest.MockedFunction<typeof uuidv4>;

describe('SampleSource', () => {
	let sampleSource: SampleSource;

	beforeEach(() => {
		jest.spyOn(console, 'error').mockImplementation(jest.fn());
		jest.spyOn(console, 'log').mockImplementation(jest.fn());
		sampleSource = new SampleSource(providerUrl);
		jest.clearAllMocks();
		mockedUuid.mockReturnValue('123e4567-e89b-12d3-a456-426614174000' as unknown as Uint8Array);
	});

	it('should fetch and parse XML data into TransactionModel objects', async () => {
		mockedAxios.get.mockResolvedValue({ data: xmlResponse });

		const transactions = await sampleSource.getTransactions();

		expect(transactions).toHaveLength(3);
		expect(transactions).toMatchSnapshot();
	});

	it('should throw an error when the XML data is invalid', async () => {
		const invalidXmlResponse = '<invalid><data></data></invalid>';
		mockedAxios.get.mockResolvedValue({ data: invalidXmlResponse });

		await expect(sampleSource.getTransactions()).rejects.toThrow(
			'Could not fetch transactions from the provider.',
		);
	});

	it('should throw an error when the provider URL is unreachable', async () => {
		mockedAxios.get.mockRejectedValue(new Error('Network Error'));

		await expect(sampleSource.getTransactions()).rejects.toThrow(
			'Could not fetch transactions from the provider.',
		);
	});
});
