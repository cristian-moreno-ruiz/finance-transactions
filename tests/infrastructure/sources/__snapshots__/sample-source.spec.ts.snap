// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleSource should fetch and parse XML data into TransactionModel objects 1`] = `
[
  {
    "amount": -100,
    "currency": "EUR",
    "date": 2024-06-30T21:00:00.000Z,
    "description": "Payment for Rent",
    "externalAccountId": 1234,
    "externalTransactionId": 291,
    "id": "123e4567-e89b-12d3-a456-426614174000",
    "source": "SAMPLE_SOURCE",
    "status": "COMPLETED",
    "title": "Direct Debit",
  },
  {
    "amount": 1000,
    "currency": "EUR",
    "date": 2024-07-31T20:00:00.000Z,
    "description": "Payment for Salary",
    "externalAccountId": 1234,
    "externalTransactionId": 1642,
    "id": "123e4567-e89b-12d3-a456-426614174000",
    "source": "SAMPLE_SOURCE",
    "status": "COMPLETED",
    "title": "Paycheck",
  },
  {
    "amount": -50,
    "currency": "EUR",
    "date": 2024-10-01T20:00:00.000Z,
    "description": "Payment for Electricity",
    "externalAccountId": 1234,
    "externalTransactionId": 1642,
    "id": "123e4567-e89b-12d3-a456-426614174000",
    "source": "SAMPLE_SOURCE",
    "status": "PENDING",
    "title": "Electricity Bill",
  },
]
`;
